### PR TITLE
Versioned search

### PR DESCRIPTION
**Pull request by riccardozambito**
Versioned search implemented.

Fix and improvements reported in PR  #1  are applied.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -63,13 +63,48 @@ module.exports = function (context, options) {
...
           id: searchDocuments.length,
           title: d.title,
           content: d.content,
-          keywords: d.keywords
-        });
+          keywords: d.keywords,
+        }
+        
+        if(version)
+          lunrAddOptions={...lunrAddOptions,...{version}};
```
Comment (davidebianchi) [inline src/index.js]:
```suggestion
        if(version) {
          lunrAddOptions={...lunrAddOptions,...{version}};
        }
```

File: src/index.js [hunk 2]
```diff
@@ -63,13 +63,48 @@ module.exports = function (context, options) {
...
+                "versions.json"
+              ),
+              "utf-8"
+            )
+          ),
+          "next",
+        ];
+
+        // Determine the version of current document
+        let version = null;
```
Comment (davidebianchi) [inline src/index.js:85]: What happen in default docs, so without version?

File: src/theme/SearchBar/index.js [hunk 3]
```diff
@@ -5,16 +5,58 @@
...
+import React, { useRef, useCallback, useState, useEffect, version} from "react";
 import classnames from "classnames";
-import { useHistory } from "@docusaurus/router";
+import { useHistory, useLocation } from "@docusaurus/router";
 import useDocusaurusContext from "@docusaurus/useDocusaurusContext";
+import useVersioning from "@theme/hooks/useVersioning";
+
+const urlMatchesPrefix = (url, prefix) => {
+  if (prefix.endsWith("/")) {
+    throw new Error(`prefix must not end with a /. This is a bug.`);
```
Comment (davidebianchi) [inline src/theme/SearchBar/index.js]:
```suggestion
    throw new Error(`prefix must not end with a /.`);
```

File: src/theme/SearchBar/index.js [hunk 4]
```diff
@@ -32,7 +74,8 @@ const Search = props => {
           // So, we can safely remove it. See https://github.com/facebook/docusaurus/issues/1828 for more details.
 
           history.push(url);
-        }
+        },
+        versionsToSearch: versionToSearch ? [versionToSearch] : null,
```
Comment (davidebianchi) [inline src/theme/SearchBar/index.js:78]: If versionToSearch is null, are searched all the versions?

File: src/theme/SearchBar/lib/lunar-search.js [hunk 5]
```diff
@@ -16,6 +16,13 @@ class LunrSearchAdapter {
             query.term(tokens, {
                 wildcard: lunr.Query.wildcard.TRAILING
             });
+
+            if(versions) {
+                query.term(versions.join(" "), {
+                    fields: ["version"],
+                    boost: 0,
```
Comment (davidebianchi) [inline src/theme/SearchBar/lib/lunar-search.js:23]: What is boost 0?

File: src/theme/SearchBar/index.js (hunk 4 above)
Comment (riccardozambito) [reply src/theme/SearchBar/index.js:78]: yes, they are

File: src/index.js (hunk 2 above)
Comment (riccardozambito) [reply src/index.js:85]: If versioning is disabled all docs will be searched.

File: src/theme/SearchBar/lib/lunar-search.js (hunk 5 above)
Comment (riccardozambito) [reply src/theme/SearchBar/lib/lunar-search.js:23]: > In multi-term searches, a single term may be important than others. For these cases Lunr supports term level boosts. Any document that matches a boosted term will get a higher relevance score, and appear higher up in the results.

https://lunrjs.com/guides/searching.html#boosts

I have lost the code that removes results with a score of 0 that occur when the docs are versioned and just the version matches. This Is the reason for set 0 as "weight" of version term.

I have added the code after these lines.

File: src/index.js (hunk 2 above)
Comment (davidebianchi) [reply src/index.js:85]: But the default version is the latest, and the common case. So if I search without select a version, I see al the versions in search?

Comment (riccardozambito) [reply src/index.js:85]: This part of the code assigns the version to each document. The version is null when docs versioning is disabled.
The code which set the version to use in searching it's [src/theme/SearchBar/index.js](src/theme/SearchBar/index.js#L49-L57) .
If you search without select version you should get only results from the last version. Now it only happens when a user searches from the homepage.

**Pull request merged**